extensions/wikia/Captcha/Captcha.class.php

```diff
@@ -0,0 +1,625 @@
+			}
+			$template->set( 'captcha',
+				"<div class='captcha'>" .
+				$this->captcha->getForm() .
+				'<p class="captchadesc" >' . $message . '</p>' .
+				"</div>\n" );
+		}
+		return true;
+	}
+
+	/**
+	 * Inject a captcha into the user login form after a failed
+	 * password attempt as a speedbump for mass attacks.
+	 *
+	 * @fixme if multiple thingies insert a header, could break
+	 *
+	 * @param \QuickTemplate $template
+	 *
+	 * @return bool whether to keep running callbacks
+	 */
+	public function injectUserLogin( &$template ) {
+		if ( $this->isBadLoginTriggered() ) {
+			$template->set( 'header',
+				"<div class='captcha'>" .
+				\F::app()->wg->Out->parse( $this->captcha->getMessage( 'badlogin' ) ) .
+				$this->captcha->getForm() .
+				"</div>\n" );
+		}
+		return true;
+	}
+
+	/**
+	 * When a bad login attempt is made, increment an expiring counter
+	 * in the memcache cloud. Later checks for this may trigger a
+	 * captcha display to prevent too many hits from the same place.
+	 *
+	 * @param \User $user
+	 * @param string $password
+	 * @param int $retval authentication return value
+	 *
+	 * @return bool whether to keep running callbacks
+	 */
+	public function triggerUserLogin( $user, $password, $retval ) {
+		$wg = \F::app()->wg;
+		if ( $retval == \LoginForm::WRONG_PASS && $wg->CaptchaTriggers['badlogin'] ) {
+			$key = $this->badLoginKey();
+			$count = $wg->Memc->get( $key );
+			if ( !$count ) {
+				$wg->Memc->add( $key, 0, self::BAD_LOGIN_TTL );
+			}
+			$wg->Memc->incr( $key );
+		}
+		return true;
+	}
+
+	/**
+	 * Check if a bad login has already been registered for this
+	 * IP address. If so, require a captcha.
+	 *
+	 * @return bool
+	 */
+	private function isBadLoginTriggered() {
+		$wg = \F::app()->wg;
+
+		if ( empty( $wg->CaptchaTriggers['badlogin'] ) ) {
+			return false;
+		}
+
+		$loginAttempts = $wg->Memc->get( $this->badLoginKey() );
+		return $loginAttempts >= self::MAX_BAD_LOGIN_ATTEMPTS;
+	}
+
+	/**
+	 * Check if the IP is allowed to skip captchas
+	 */
+	public function isIPWhitelisted() {
+		$wg = \F::app()->wg;
+
+		if ( $wg->CaptchaWhitelistIP ) {
+			$wg = \F::app()->wg;
+
+			$ip = $wg->Request->getIP();
+
+			foreach ( $wg->CaptchaWhitelistIP as $range ) {
+				if ( \IP::isInRange( $ip, $range ) ) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Internal cache key for badlogin checks.
+	 *
+	 * @return string
+	 */
+	private function badLoginKey() {
+		$ip = \F::app()->wg->Request->getIP();
+		return wfMemcKey( 'captcha', 'badlogin', 'ip', $ip );
+	}
+
+	/**
+	 * @param \EditPage $editPage
+	 * @param string $action (edit/create/addurl...)
+	 *
+	 * @return bool true if action triggers captcha on editPage's namespace
+	 */
+	public function captchaTriggers( &$editPage, $action ) {
+		$wg = \F::app()->wg;
+		$ns = $editPage->mTitle->getNamespace();
+
+		// Special config for this NS?
+		if ( isset( $wg->CaptchaTriggersOnNamespace[$ns][$action] ) )
```

Comment:
Do we have a convention in PHP to use curly braces even if the expression is only one line? I think it's best practice to include them for clarity.